File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] arr) {
        int leftmax=0,rightmax=0,total=0;
        int n=arr.length;
        int l=0,r=n-1;
        while(l<r){
            if(arr[l]<=arr[r]){
                if(leftmax>arr[l]) total+=leftmax-arr[l];
                else leftmax=arr[l];
                l=l+1;
            }
            else{
                if(rightmax>arr[r]){
                    total+=rightmax-arr[r];
                }
                else rightmax=arr[r];
                r=r-1;
            }
        }
        return total;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, arr: List[int]) -> int:
        n=len(arr)
        leftmax=rightmax=total=0
        l,r=0,n-1
        while(l<r):
            if arr[l]<=arr[r]:
                if leftmax>arr[l]:
                    total+=leftmax-arr[l]
                else:
                    leftmax=arr[l]
                l+=1
            else:
                if rightmax>arr[r]:
                    total+=rightmax-arr[r]
                else:
                    rightmax=arr[r]
                r-=1
        return total